bindings/python/pytpm2tss/__init__.py

```diff
@@ -0,0 +1,2533 @@
+"""
+SPDX-License-Identifier: BSD-2
+
+Copyright 2018, Fraunhofer SIT
+All rights reserved.
+"""
+
+from . import headers
+from cffi import FFI
+
+_ffi = FFI()
+
+_ffi.cdef(headers.tss2_common_h)
+_ffi.cdef(headers.tss2_tpm2_types_h)
+_ffi.cdef(headers.tss2_tcti_h)
+_ffi.cdef(headers.tss2_esys_h)
+
+_lib = _ffi.verify('#include <tss2/tss2_esys.h>', libraries=['tss2-esys'])
+
+### internal ###
+
+def _chkrc(rc):
+    if rc != 0:
+        print("ERROR occured: 0x%08x" % rc)
+        raise Exception(rc >> 16, rc & 0xffff)
+
+#### Utilities ####
+
+def TPM2B_unpack(x):
+    return _ffi.unpack(x.buffer, x.size)
+
+def TPM2B_pack(x, t='DIGEST'):
+    if t[0:6] == "TPM2B_":
+        t=t[6:]
+    r = _ffi.new("TPM2B_%s *" % t)
+    r.size = len(x)
+    for i in range(r.size):
+        r.buffer[i] = x[i]
+    return r
+
+#### ESYS_TR classes ####
+
+class EsysTr():
+    def __init__(self, ctx, handle):
+        self.ctx = ctx
+        self.handle = handle
+        if ctx:
+            self.active = True
+        else:
+            self.active = False
+
+    def __del__(self):
+        if self.active:
+            _lib.Esys_FlushContext(self.ctx.ctx, self.handle)
```

Comment:
The "allocation" happens in the `StartAuthSession` and then the allocated TPM resource is put into this EsysTr container class for resource management.

Ok, that missing guarantee an issue. I guess I'll add a `def flush(self):` for explicit ops then. But for the sloppy users or for convenience I'd like to still keep this automatic flushing around.

Is there really no way in CPython to force an invocation of a garbage collector or any other kind of trapping on application exit / error ?

(Of course this would explain the pain that TCP ports are always kept busy after python applications crash)

Reply:
Since I'm such a novice, could you give me some (more specific) code to insert here ?

Reply:
Hmmm, I think this is the correct context management using approach. Although I'm fuzzy on when `Esys_FlushContext` should be called so I don't know if this is right.

```python
# __enter__ method (nop) called when using `with`
with e.tr.OWNER as owner, e.tr.PASSWORD as password:
    owner.setAuth("")
    x, _, _, _, _ = e.CreatePrimary(owner,
                            inSensitive, inPublic, outsideInfo, creationPCR,
                            session1=password)
    # __exit__ method, lib.Esys_FlushContext(self.ctx.ctx, self.handle) called
    # when code exits the with block
```

Patch:

```
$ patch -p1 << 'EOF'
diff --git a/bindings/python/pytpm2tss/__init__.py b/bindings/python/pytpm2tss/__init__.py
index b7f5b12f..9de2a123 100644
--- a/bindings/python/pytpm2tss/__init__.py
+++ b/bindings/python/pytpm2tss/__init__.py
@@ -34,15 +34,12 @@ class EsysTr:
     def __init__(self, ctx, handle):
         self.ctx = ctx
         self.handle = handle
-        if ctx:
-            self.active = True
-        else:
-            self.active = False
 
-    def __del__(self):
-        if self.active:
-            lib.Esys_FlushContext(self.ctx.ctx, self.handle)
-        super().__del__(self)
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc, exc_tb):
+        lib.Esys_FlushContext(self.ctx.ctx, self.handle)
 
     def setAuth(self, auth):
         if type(auth) is str:
diff --git a/test/bindings/pytpm2tss b/test/bindings/pytpm2tss
index cfcef5bd..c05b6f02 100755
--- a/test/bindings/pytpm2tss
+++ b/test/bindings/pytpm2tss
@@ -65,11 +65,11 @@ inPublic.publicArea.parameters.eccDetail.symmetric.algorithm = TPM2_ALG.NULL
 inPublic.publicArea.parameters.eccDetail.kdf.scheme = TPM2_ALG.NULL
 inPublic.publicArea.parameters.eccDetail.curveID = TPM2_ECC.NIST_P256
 
-e.tr.OWNER.setAuth("")
-
-x, _, _, _, _ = e.CreatePrimary(e.tr.OWNER, 
-                        inSensitive, inPublic, outsideInfo, creationPCR,
-                        session1=e.tr.PASSWORD)
+with e.tr.OWNER as owner, e.tr.PASSWORD as password:
+    owner.setAuth("")
+    x, _, _, _, _ = e.CreatePrimary(owner,
+                            inSensitive, inPublic, outsideInfo, creationPCR,
+                            session1=password)
 
 
 print("OK")
EOF
```